pycmds/_record.py

```diff
@@ -352,160 +267,42 @@ def q(method, inputs=[]):
-
-    def close(self):
-        # TODO:
-        if False:
-            # stop looping
-            self.set_freerun(False)
-            self.wait_until_done()
-            # TODO: log
-            if False:
-                if g.debug.read():
-                    print("daq shutting down")
-                g.logger.log("info", "DAQ shutdown")
-            # shutdown driver
-            q.push("shutdown")
-            self.wait_until_done()
-            self.thread.quit()
-        else:
-            self.thread.quit()
-
-    def get_headers(self):
-        out = collections.OrderedDict()
-        out["shots"] = self.nshots.read()
-        return out
-
-    def give_widget(self, widget):
-        self.widget = widget
-        self.gui.create_frame(widget)
-
-    def initialize(self):
-        self.wait_until_still()
-        self.freerun.updated.connect(self.on_freerun_updated)
-        self.update_ui.emit()
-        self.driver.update_ui.connect(self.on_driver_update_ui)
-        self.settings_updated.emit()
-
-    def load_settings(self, aqn):
-        pass
-
-    def measure(self):
-        self.q.push("measure")
-
-    def on_driver_update_ui(self):
-        self.update_ui.emit()
-
-    def on_freerun_updated(self):
-        self.q.push("loop")
-
-    def set_freerun(self, state):
-        self.freerun.write(state)
-        self.on_freerun_updated()
-        self.settings_updated.emit()  # TODO: should probably remove this
-
-    def wait_until_still(self):
-        while self.busy.read():
-            self.busy.wait_for_update()
-
-
-# --- driver --------------------------------------------------------------------------------------
-
-
-class Driver(pc.Driver):
-    settings_updated = QtCore.Signal()
-    running = False
-
-    def __init__(self, device):
-        pc.Driver.__init__(self)
-        # attributes
-        self.name = "Virtual"
-        self.enqueued = device.enqueued
-        self.busy = device.busy
-        self.freerun = device.freerun
-        self.data = device.data
-        self.shape = device.shape
-        self.measure_time = device.measure_time
-        self.thread = device.thread
-
-    def initialize(self):
-        self.measure()
-
-    def loop(self):
-        while self.freerun.read() and not self.enqueued.read():
-            self.measure()
-            self.busy.write(False)
-
-    def measure(self):
-        timer = wt.kit.Timer(verbose=False)
-        with timer:
-            time.sleep(0.1)
-            out_names = ["channel_%i" % i for i in range(5)]
-            signed = [not i % 2 for i in range(5)]
-            out = np.random.standard_normal(len(out_names))
-            self.data.write_properties(self.shape, out_names, out, signed)
-        self.measure_time.write(timer.interval)
-        self.update_ui.emit()
-
-    def shutdown(self):
-        pass
-
-
-# --- gui -----------------------------------------------------------------------------------------
-
-
 class Widget(QtWidgets.QWidget):
```

Comment:
There were _two_ Widget classes previously, this looks like it is largely the _second_ one (the one that was _actually_ used previously), but at the position of the first, is this correct?